sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -578,12 +586,51 @@ void validate(PipelineOptions options) {
               "The following transforms do not have stable unique names: {}",
               Joiner.on(", ").join(unstableNames));
           break;
-        case ERROR:
+        case ERROR: // be very verbose here since it will just fail the execution
           throw new IllegalStateException(
               String.format(
                   "Pipeline update will not be possible"
                       + " because the following transforms do not have stable unique names: %s.",
-                  Joiner.on(", ").join(unstableNames)));
+                  Joiner.on(", ").join(unstableNames)) + "\n\n"
+                      + "Conflicting instances:\n"
+                      + Joiner.on("\n").join(Iterables.transform(unstableNames,
```

Comment:
Hmm I'm having a hard time following this code for two reasons:
- I don't remember in what format are the unstable names here. It would help to add some comments with examples of intermediate values at each step
- Java7 syntax makes the functional programming style here very verbose, I think this would be more readable if refactored into simple loops and static utility functions (ideally with their own unit tests)